Improve @AuthenticationPrincipal meta-annotations

kse-music opened the pull request:
Closes gh-15286

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

jzheaux commented on messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -83,15 +94,20 @@
+
 	private ExpressionParser parser = new SpelExpressionParser();
 
+	private PrePostTemplateDefaults defaults = new PrePostTemplateDefaults();
```
I think we'll want to introduce a new class, maybe `AuthenticationPrincipalTemplateDefaults`. By the name and JavaDoc `PrePostTemplateDefaults` is intended for the pre/post annotations.

kse-music replied:
Done

jzheaux commented on messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -133,26 +149,63 @@ public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy secur
+	 * @since 6.4
+	 */
+	public void setTemplateDefaults(@Nullable PrePostTemplateDefaults defaults) {
+		if (defaults != null) {
```
Please instead don't allow null values. We don't want methods to silently not do what they say they will do.

kse-music replied:
Done

jzheaux commented on messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -133,26 +149,63 @@ public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy secur
-			}
-		}
-		return null;
+		return (T) this.cachedAttributes.computeIfAbsent(parameter, (k) -> {
```
Let's consider turning `AuthorizationAnnotationUtils` into an API that both the argument resolver and method security can use. Do you have time for that? If not, I can create a separate ticket for it.

kse-music replied:
it's my pleasure

jzheaux reviewed (changes requested):
Thanks for contributing this, @kse-music! I've left some initial feedback inline.

kse-music commented:
Thanks,@jzheaux ,all feedback is done

jzheaux commented:
Thanks, @kse-music! Given that this needs to be placed in the other `AuthenticationPrincipalArgumentResolver`s, I extracted the logic in `AuthorizationAnnotationUtils` into an API. Please see if it works for you. If so, please also update this PR to change the other non-deprecated instances of `AuthenticationPrincipalArgumentResolver` in the project.

kse-music commented:
> Thanks, @kse-music! Given that this needs to be placed in the other `AuthenticationPrincipalArgumentResolver`s, I extracted the logic in `AuthorizationAnnotationUtils` into an API. Please see if it works for you. If so, please also update this PR to change the other non-deprecated instances of `AuthenticationPrincipalArgumentResolver` in the project.

All non-deprecated instances of AuthenticationPrincipalArgumentResolver in the project has done.

jzheaux commented:
Thanks, @kse-music, this is taking shape very nicely.

I believe our last step is to make this simple to configure by picking up the `AnnotationTemplateExpressionDefaults` bean in configuration.

I've added [documentation](https://github.com/spring-projects/spring-security/pull/15344/files#diff-cf8343e91909a111571bdcae3d007cb61824793545041ec143247ce3e69c1c85) so you can see the expectation that someone should be able to do:

```java
@Bean
public AnnotationTemplateExpressionDefaults templateExpressionDefaults() {
    return new AnnotationTemplateExpressionDefaults();
}
```

And now they will be able to use placeholders in any `@AuthenticationPrincipal` meta-annotation definition.

Are you able to add this to the following `spring-security-config` classes: `WebMvcSecurityConfiguration`, `ServerHttpSecurityConfiguration`, and `WebSocketMessageBrokerSecurityConfiguration`? It's a little harder, however, if you are able, it would be great to add to `WebSocketMessageBrokerSecurityBeanDefinitionParser`, too.

kse-music commented:
@jzheaux ,I tried to add it

kse-music commented:
@jzheaux I have added the code, please help to review it.Thank you

jzheaux commented:
Thanks for all your help with this PR, @kse-music! I added a couple of commits to add more testing and also to deprecate `PrePostTemplateDefaults`. It's now merged into `main`.

jzheaux commented:
If you are interested, it would be nice to do the same with `@CurrentSecurityContext`. I've added https://github.com/spring-projects/spring-security/issues/15551 to address that. That PR should go a lot more quickly.

kse-music commented:
> If you are interested, it would be nice to do the same with `@CurrentSecurityContext`. I've added #15551 to address that. That PR should go a lot more quickly.

Yes,I have added PR #15553
